feat: make the android XXXRes annotations aliases to checker annotations and added tests

aagarwal1012 opened the pull request.

aagarwal1012 committed:
feat: make the android XXXRes annotations aliases to checker annotations

aagarwal1012 committed:
feat: added androidx annotation alias

aagarwal1012 committed:
feat: added android annotation alias

aagarwal1012 committed:
ptch: added tests for aliases

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -219,11 +228,52 @@
             AnnotationBuilder.fromClass(elements, XmlContainer.class);
 
+    /**
+     * Aliases for Android @{@code XXXRes} annotations.
```
These are not aliases and not annotations. This field contains package prefixes for the actual annotations.
Please pick a more suitable name for the field and describe it correctly.

aagarwal1012 replied:
Fixed.

wmdietl commented on tests/androidresource/src/AndroidAliasesTest.java:
```diff
@@ -0,0 +1,32 @@
+    @AnyRes int any_res;
+    @RawRes int raw_res;
+
+    @FractionRes int fraction_res;
```
Instead of 5 different XXXRes annotations, I would just use two - no point in adding all the other files just for this test.

aagarwal1012 replied:
Fixed.

wmdietl commented on tests/androidresource/src/AndroidAliasesTest.java:
```diff
@@ -0,0 +1,32 @@
+        string_res = R.string.abc_action_bar_home_description;
+
+        attr_res = R.attr.actionBarDivider;
+        xml_res = R.xml.file_paths;
```
You should add failing tests, to actually show that the annotations are interpreted by the type system. At the moment you might not have done anything and the test would continue to work.

aagarwal1012 replied:
Fixed.

aagarwal1012 committed:
chor: fixed documentation

aagarwal1012 committed:
ptch: revmoved some test annotations and added some failing tests

aagarwal1012 committed:
conf: updated CF version in build.gradle

aagarwal1012 commented:
@wmdietl, 
I have added some failing tests but they passed successfully. Then I checked the `addAliasedAnnotation` method is not working. I got the log below of a test case:

**Test case:**
```java
@android.annotations.RawRes int raw_res = R.raw.raw_res;
``` 
**Log:**
```java
AndroidResourceVisitor about to test whether actual is a subtype of expected (line  25): MEMBER_SELECT R.raw.raw_res
         actual: INT @RawRes int
       expected: INT @ResourceTop int
     success: actual is subtype of expected (line  25): MEMBER_SELECT R.raw.raw_res
         actual: INT @RawRes int
       expected: INT @ResourceTop int

```
Now, what to do?

aagarwal1012 committed:
Merge remote-tracking branch 'origin/master' into develop

wmdietl commented:
Your comment says "I have added some failing tests but they passed successfully. " but `AndroidAliasesTest.java` still contains no expected errors and the test suite is passing.
Please add expected error markers so that it is clear where you expect errors and to make sure the test suite fails if they are not issued.

aagarwal1012 committed:
ptch: added expected error message in the AndroidAliasesTest

aagarwal1012 commented:
@wmdietl, 
I have updated the branch and added the _expected error messages_ with the tests. Now, you can see that Travis fails. Please once again recheck my previous comment.

aagarwal1012 committed:
bfix: changed package name of android test annotations

aagarwal1012 committed:
ptch: updated error messages in android aliases test

aagarwal1012 commented:
OMG! such a small mistake and it took hours to figure out, finally solved. Actually, I made the android test annotation's package name to `android.annotations` but it should be `android.annotation`, such a small mistake of an alphabet make to read the whole code thrice. 
Sorry Werner, now please review the PR.

aagarwal1012 committed:
Merge remote-tracking branch 'origin/master' into develop

aagarwal1012 committed:
feat: updated makeAndroidAnnotationsAliasesToChecker method

wmdietl commented on src/test/java/android/annotation/AnyRes.java:
```diff
@@ -0,0 +1,13 @@
+package android.annotation;
```
I would add some documentation, maybe both on the individual annotations and in the README file, explaining that these annotations are only included for testing purposes.
For example, in the CF, such annotations are in a separate `checker/src/testannotations/` directory:

https://github.com/typetools/checker-framework/blob/master/checker/src/testannotations/java/android/annotation/NonNull.java

And the beginning of that file points to the canonical version of the annotation.

aagarwal1012 replied:
Fixed.

wmdietl commented on tests/androidresource/src/AndroidAliasesTest.java:
```diff
@@ -0,0 +1,64 @@
+
+public class AndroidAliasesTest {
+
+    @AnyRes
```
Do not add a line-break between the type annotation and the type it annotates. It is much nicer to read as `@AnyRes int any_res`.

aagarwal1012 replied:
Fixed.

wmdietl commented on tests/androidresource/src/AndroidAliasesTest.java:
```diff
@@ -0,0 +1,64 @@
+import androidx.annotation.StringRes;
+import android.annotation.AnyRes;
+import android.annotation.RawRes;
```
You should also include a few tests that show that the android annotations are aliases to your annotations, that is, that you can assign a `android.annotation.RawRes` variable to your version of `@RawRes`.

aagarwal1012 replied:
Fixed.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -259,11 +267,56 @@
+    private void makeAndroidAnnotationsAliasesToChecker() {
+
+        ANDROID_PACKAGE_PREFIXES.forEach((annotation) -> {
+            addAliasedAnnotation(annotation + "AnyRes", ANY_RES);
```
I wonder if you can use another lambda to further reduce the duplication in this code - map over (String, Annotation) and map to the expression here.

aagarwal1012 replied:
Fixed.

wmdietl reviewed (commented):
This looks nice! A few small comments to further improve it.

aagarwal1012 committed:
chor: added nested lambda in makeAndroidAnnotationsAliasesToChecker method

aagarwal1012 committed:
docs: added docs for android test annotations

aagarwal1012 committed:
chor: removed line-breaks between annotations

aagarwal1012 committed:
ptch: added tests

Tests added to show that the android annotations are aliases to the checker annotations.

aagarwal1012 commented:
@wmdietl, 
I have made the required changes that you suggested to me.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -5,6 +5,7 @@
 import org.checkerframework.checker.androidresource.qual.ResourceTop;
 import org.checkerframework.checker.androidresource.qual.container.*;
 import org.checkerframework.checker.androidresource.qual.res.*;
+import org.checkerframework.com.google.common.collect.ImmutableMap;
```
I think in general we have not been using these Guava dependencies and it just happens to be available because of dependencies from AFU pulling them in.
Would it be much longer if you used just standard Java APIs?

aagarwal1012 replied:
Fixed.

wmdietl reviewed (approved):
One small comment, otherwise looks good!

aagarwal1012 committed:
chor: replaced Guava ImmutableMap to java HashMap

aagarwal1012 commented:
@wmdietl,
I have replaced the Guava `ImmutableMap` with java `HashMap`.

aagarwal1012 commented:
Now, merging this PR.
